daemon/rrd-update.c

```diff
@@ -240,37 +243,74 @@ void rb_rrd_update(rb_poller *poll)
-                log_errorx("couldn't create directory for raw file: %s : %s",
-                            path,  strerror(errno));
+                log_errorx("raw file: %s: mkdir: %s", path, strerror(errno));
                 free(parent);
-                /* Try the next item */
-                continue;
+                break; /* next raw file */
             }
             free(parent);
-        
-            fp = fopen(path, "a");
-            if(fp == NULL)
+
+            fd = open(path, O_WRONLY|O_APPEND|O_CREAT, 0644);
+            if(fd == -1)
             {
-                log_errorx("couldn't open raw file: %s for writing : %s",
-                            path,  strerror(errno));
-                /* Try the next item */
-                continue;
+                log_errorx("raw file: %s: open: %s", path, strerror(errno));
+                break; /* next raw file */
             }
 
-            /* Write item record to raw file */
-            fprintf(fp, "%" PRId64 "\t", time);
-            if(item->reference)
-                fprintf(fp, "%s", item->reference);
-            else
-                fprintf(fp, "%s", item->field);
-            fprintf(fp, "\t");
-            if(item->vtype == VALUE_REAL)
-                fprintf(fp, "%" PRId64, item->v.i_value);
-            else if(item->vtype == VALUE_FLOAT)
-                fprintf(fp, "%.4lf", item->v.f_value);
-            fprintf(fp, "\n");
-
-            fclose(fp);
+            write_sample(fd, path, &time, item);
+
+            if (close(fd) == -1)
+            {
+                log_errorx("raw file: %s: close: %s", path, strerror(errno));
+            }
         }
     }
 }
+
+static void
+write_sample(int fd, const char* path, const time_t *time, const rb_item *item)
+{
+    char buf[RAW_BUFLEN];
+    ssize_t nw;
+    int n;
+
+    switch (item->vtype) {
+    case VALUE_REAL:
+        n = snprintf(buf, sizeof(buf), "%"PRId64"\t%s\t%"PRId64"\n",
+          *time,
+          (item->reference ? item->reference : item->field),
+          item->v.i_value);
+        break;
+
+    case VALUE_FLOAT:
+        n = snprintf(buf, sizeof(buf), "%"PRId64"\t%s\t%.4lf\n",
+          *time,
+          (item->reference ? item->reference : item->field),
+          item->v.f_value);
+        break;
+
+    case VALUE_UNSET:
+        n = snprintf(buf, sizeof(buf), "%"PRId64"\t%s\t\n",
+          *time,
+          (item->reference ? item->reference : item->field));
+        break;
+
+    default:
+        log_errorx("raw file: %s: unknown sample value type: %d", path, item->vtype);
+        return;
+    }
+
+    if (n == -1) {
+        log_errorx("raw file: %s: snprintf: %s", path, strerror(errno));
+        return;
+    }
+
+    if (n >= sizeof(buf)) {
+        log_errorx("raw file: %s: truncated sample string: required: %d", path, n);
+        return;
+    }
+
+    if ((nw = write(fd, buf, n)) == -1 || nw != n) {
+        log_errorx("raw file: %s: write: %s", path, strerror(errno));
```

Comment:
Sorry, this may sound pedantic, but errno will be set only if `-1` is returned.  For the case when only partial buffer is written (so likely only in an out-of-diskspace situation), errno would be unset (according to my reading of man `write(2)`).

Reply:
(*path to write_sample cleared up over slack) 
heh, thanks for picking that one -- strerror(0) (== unset errno) gives me "Undefined error: 0", which is correct: it didn't write all the bytes, which is an error, but we don't know what the error is. But I do see in intro(2) openbsd that "[errno] should only be examined after an error." So I will change this. 

In my defense, I was supported by OpenBSD dd, which loops over 
if ((n = write(out.fd, out.db, out.dbsz)) != out.dbsz) 
    err(1, "%s", out.name);
... but this is dubious to me now in light of the above.